# Lab 4, Exercise 1

In [1]:
import numpy as np
import pandas as pd

## Load data 

Here is an example of the first couple rows from the data:

| id | dur       | proto | service | state | spkts | dpkts | sbytes | dbytes | rate        | sttl | dttl | sload     | dload | sloss | dloss | sinpkt | dinpkt | sjit | djit | swin | stcpb | dtcpb | dwin | tcprtt | synack | ackdat | smean | dmean | trans\_depth | response\_body\_len | ct\_srv\_src | ct\_state\_ttl | ct\_dst\_ltm | ct\_src\_dport\_ltm | ct\_dst\_sport\_ltm | ct\_dst\_src\_ltm | is\_ftp\_login | ct\_ftp\_cmd | ct\_flw\_http\_mthd | ct\_src\_ltm | ct\_srv\_dst | is\_sm\_ips\_ports | attack\_cat | label |
|----|-----------|-------|---------|-------|-------|-------|--------|--------|-------------|------|------|-----------|-------|-------|-------|--------|--------|------|------|------|-------|-------|------|--------|--------|--------|-------|-------|--------------|---------------------|--------------|----------------|--------------|---------------------|---------------------|-------------------|----------------|--------------|---------------------|--------------|--------------|--------------------|-------------|-------|
| 1  | 0\.000011 | udp   | \-      | INT   | 2     | 0     | 496    | 0      | 90909\.0902 | 254  | 0    | 180363632 | 0     | 0     | 0     | 0\.011 | 0      | 0    | 0    | 0    | 0     | 0     | 0    | 0      | 0      | 0      | 248   | 0     | 0            | 0                   | 2            | 2              | 1            | 1                   | 1                   | 2                 | 0              | 0            | 0                   | 1            | 2            | 0                  | Normal      | 0     |


In [2]:
# Load the training and test data corresponding to exercise 1
# Create two separate pandas dataframes for the training and test data
# For each dataframe, import the following CSV data
# training set: 'data/exercise1/UNSW_NB15_training-set.csv'
# test set: 'data/exercise1/UNSW_NB15_testing-set.csv'

# CODE HERE
train = pd.read_csv('data/exercise1/UNSW_NB15_training-set.csv',index_col='id')
test = pd.read_csv('data/exercise1/UNSW_NB15_testing-set.csv',index_col='id')

### Part 1 
Keep the two dataframes separate and create train/test data and labels.  This will be used to experiment with the case where there are different types of activities in the training versus test set.

In [3]:
# Do all the following operations in the cell on both the dataframes separately
# 1) Keep only the datapoints where the 'attack_cat' column is equal to either 'Normal' or 'Fuzzers'
# CODE HERE
train_data = train.loc[(train['attack_cat'] == 'Normal') | (train['attack_cat'] == 'Fuzzers')]
test_data = test.loc[(test['attack_cat'] == 'Normal') | (test['attack_cat'] == 'Fuzzers')]

# 2) Get the labels from the dataframe (i.e., the values in the 'attack_cat' column)
# CODE HERE
train_labels = train_data['attack_cat']
test_labels = test_data['attack_cat']

# 3) Keep only the features we care about for this experiment.
# We only care about the numerical features between column 'spkts' and 'is_sm_ips_ports' (inclusive)
# CODE HERE
train_data = train_data.iloc[:,4:42]
test_data = test_data.iloc[:,4:42]

# You should now have four inputs usable for scikit-learn:
# training data -> train_data
# training labels -> train_labels
# test data -> test_data
# test labels -> test_labels
# Hint: You may have to add some minor code in the above to get the data ready for scikit-learn

### Part 2 
Create a new training/test split by combining the dataframes into one.  Then split the dataframe randomly into train/test data and labels.  This will be used to experiment with the case where there are largely the same types of activities in the training and test set.

In [4]:
# Combine the dataframes into a single dataframe, then do the following
# CODE HERE
com = pd.concat([train, test], ignore_index=True)

# 1) Keep only the datapoints where the 'attack_cat' column is equal to either 'Normal' or 'Fuzzers'
# CODE HERE
com_data = com.loc[(com['attack_cat'] == 'Normal') | (com['attack_cat'] == 'Fuzzers')]

# 2) Get the labels from the dataframe (i.e., the values in the 'attack_cat' column)
# CODE HERE
com_labels = com_data['attack_cat']

# 3) Keep only features we care about for this experiment.
# We only care about the numerical features between column 'spkts' and 'is_sm_ips_ports' (inclusive)
# CODE HERE
com_data = com_data.iloc[:,4:42]

# 4) Create a random split; put 50% of the data into the training set and the other 50% into the test set
# Use scikit-learn's 'train_test_split'
# Hint: You may have to add some minor code in the above to get the data ready for scikit-learn's 'train_test_split'
# CODE HERE
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(com_data, com_labels, test_size=0.5, random_state=42)

# You should now have four inputs usable for scikit-learn that are distinct from the inputs you created in Part 1.
# The inputs should correspond to:
# training data -> X_train
# training labels -> y_train
# test data -> X_test
# test labels -> y_test

## Train a classifier

In [5]:
# For each of the train/test splits, create a separate random forest with default sklearn parameters
# Hint: Ignoring import statements, each random forest can be created in a single line of code
# CODE HERE
from sklearn.ensemble import RandomForestClassifier
# part1
p1_class = RandomForestClassifier(random_state=42).fit(train_data, train_labels)
# part2
p2_class = RandomForestClassifier(random_state=42).fit(X_train, y_train)

/home/sean/Documents/EE379K/EE379K-Lab4/env/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/home/sean/Documents/EE379K/EE379K-Lab4/env/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


## Inference and results

In [6]:
# For each of the train/test splits and associated random forest, do the following:

# 1) Predict labels on the training data
# CODE HERE
# part1
p1_train_preds = p1_class.predict(train_data)

# part2
p2_train_preds = p2_class.predict(X_train)

# 2) Print metrics on the training data; use sklearn's implementation of precision, recall, f1, and accuracy
# CODE HERE
from sklearn import metrics
# part1
p1_train_pre = metrics.precision_score(train_labels, p1_train_preds, pos_label='Fuzzers')
p1_train_re = metrics.recall_score(train_labels, p1_train_preds, pos_label='Fuzzers')
p1_train_f1 = metrics.f1_score(train_labels, p1_train_preds, pos_label='Fuzzers')
p1_train_acc = metrics.accuracy_score(train_labels, p1_train_preds)
# part2
p2_train_pre = metrics.precision_score(y_train, p2_train_preds, pos_label='Fuzzers')
p2_train_re = metrics.recall_score(y_train, p2_train_preds, pos_label='Fuzzers')
p2_train_f1 = metrics.f1_score(y_train, p2_train_preds, pos_label='Fuzzers')
p2_train_acc = metrics.accuracy_score(y_train, p2_train_preds)

print('TRAIN PREDS ' + 'part1'           + '\t\t' + 'part2')
print('precision = ' + str(p1_train_pre) + '\t' + str(p2_train_pre))
print('   recall = ' + str(p1_train_re)  + '\t' + str(p2_train_re))
print('f1measure = ' + str(p1_train_f1)  + '\t' + str(p2_train_f1))
print(' accuracy = ' + str(p1_train_acc) + '\t' + str(p2_train_acc))
print('\n')

# 3) Predict labels on the test data
# CODE HERE
# part1
p1_test_preds = p1_class.predict(test_data)
# part2
p2_test_preds = p2_class.predict(X_test)

# 4) Print metrics on the test data; again, use precision, recall, f1, and accuracy
# CODE HERE
# part1
p1_test_pre = metrics.precision_score(test_labels, p1_test_preds, pos_label='Fuzzers')
p1_test_re = metrics.recall_score(test_labels, p1_test_preds, pos_label='Fuzzers')
p1_test_f1 = metrics.f1_score(test_labels, p1_test_preds, pos_label='Fuzzers')
p1_test_acc = metrics.accuracy_score(test_labels, p1_test_preds)
# part2
p2_test_pre = metrics.precision_score(y_test, p2_test_preds, pos_label='Fuzzers')
p2_test_re = metrics.recall_score(y_test, p2_test_preds, pos_label='Fuzzers')
p2_test_f1 = metrics.f1_score(y_test, p2_test_preds, pos_label='Fuzzers')
p2_test_acc = metrics.accuracy_score(y_test, p2_test_preds)

print('TEST PREDS  ' + 'part1'          + '\t\t' + 'part2')
print('precision = ' + str(p1_test_pre) + '\t' + str(p2_test_pre))
print('   recall = ' + str(p1_test_re)  + '\t' + str(p2_test_re))
print('f1measure = ' + str(p1_test_f1)  + '\t' + str(p2_test_f1))
print(' accuracy = ' + str(p1_test_acc) + '\t' + str(p2_test_acc))
print('\n')

TRAIN PREDS part1		part2
precision = 0.9955416116248349	0.9736606050721683
   recall = 0.9945562520620257	0.9848234905971627
f1measure = 0.995048687902294	0.9792102349612498
 accuracy = 0.9986066601644141	0.9913515173225526


TEST PREDS  part1		part2
precision = 0.842185417856463	0.7203955112610845
   recall = 0.2432908051033876	0.7573007754495957
f1measure = 0.3775227204847037	0.7383872913734163
 accuracy = 0.8033403429310902	0.8890367261996145




## Questions:

1) For results using Part 1 data, what is the precision and recall?

The precision and recall of the predicted training labels are about 0.9955 and 0.9946, respectively. On the other hand, the precision and recall scores of the predicted testing labels are about 0.8422 and 0.2433, respectively.

2) For results using Part 1 data, describe the difference in the results on the training and test data. What does this signify? 

The results on the training data were all above 0.99 while the results on the test data were much lower for recall and f1 measure. The decrease of each score in general is likely since the test data has entries that are dissimilar to the entries in the training data. In particular, the low recall score in the test predictions for part 1 is likely due to a much greater number of false negatives than false positives, since the only difference in the formulas for precision and recall is false positives and false negatives, respectively. This means that in the test data, the model trained on the training data had a significant number of predictions of 'Normal' attacks that were actually 'Fuzzers' (false negatives), in comparison to the number of false positives, where 'Fuzzers' was predicted, but the true label was 'Normal.'

3) What changes in the results on the test data once you combine the data for Part 2? Does this produce a better classifier? Why or why not?

The test predictions have a much more balanced spread of scores since the classifier was exposed to data from the test set while training. Overall, this produces a better classifier, since accuracy (a ratio of the correctly predicted labels to the total number of data points) has gone up, despite precision going down. This simply means that in part 2, the number of false positives out of all the positive label predictions is higher compared to the ratio in part 1. In this case, if the organization is worried about fuzzing attacks, then it might be better to have more false positives and be more wary. Additionally, the recall score went up in part 2, meaning that there are relatively less false negatives in the predictions than in part 1. In other words, there are less occurrences of an attack being labeled as 'Normal' when it actually is 'Fuzzers.' Again, this is a good thing when the goal is to separate fuzzing attacks from normal attacks.